### ISSUE TITLE
OpenGraph and Twitter card enhancements

### ISSUE THREAD

**Issue by max-arnold** [getnikola/nikola#1488]
After trying new `previewimage` meta tag with Nikola 7.2, I'd like to suggest some improvements.
- [x] Make final rendered template url absolute (`<meta property="og:image">` tag). All examples I've seen always use absolute url and Twitter validation tool fails on relative one.
- [x] Add support for https://dev.twitter.com/cards/types/summary-large-image card (document that it is actually possible to change default card type using TWITTER_CARD = {'card': '...'} configuration parameter).
- [x] ~~Add support for `twitter:title`, `twitter:description`, `twitter:image` and `twitter:image:src` tags (for both summary and summary_large_image)~~
- [x] ~~There is also `og:locale` and `og:locale:alternate` meta tags which can be used to support multilanguage posts, but I'm not sure it is possible to provide appropriate locale names (Nikola uses 2-letter language codes).~~

Below is the partial patch to add necessary markup (no doc patch, no absolute url patch):

``` diff
diff --git a/nikola/data/themes/base-jinja/templates/post_helper.tmpl b/nikola/data/themes/base-jinja/templates/post_helper.tmpl
index bc6cab2..45471da 100644
--- a/nikola/data/themes/base-jinja/templates/post_helper.tmpl
+++ b/nikola/data/themes/base-jinja/templates/post_helper.tmpl
@@ -79,6 +79,16 @@
     {% elif 'creator' in twitter_card %}
     <meta name="twitter:creator" content="{{ twitter_card['creator'] }}">
     {% endif %}
+    <meta property="twitter:title" content="{{ post.title()[:70]|e }}">
+    {% if post.description() %}
+    <meta property="twitter:description" content="{{ post.description()[:200]|e }}">
+    {% else %}
+    <meta property="twitter:description" content="{{ post.text(strip_html=True)[:200]|e }}">
+    {% endif %}
+    {% if post.previewimage %}
+    <meta property="twitter:image" content="{{ url_replacer(permalink, post.previewimage, lang) }}">
+    <meta property="twitter:image:src" content="{{ url_replacer(permalink, post.previewimage, lang) }}">
+    {% endif %}
 {% endif %}
 {% endmacro %}

diff --git a/nikola/data/themes/base/templates/post_helper.tmpl b/nikola/data/themes/base/templates/post_helper.tmpl
index 9ba58d1..484df75 100644
--- a/nikola/data/themes/base/templates/post_helper.tmpl
+++ b/nikola/data/themes/base/templates/post_helper.tmpl
@@ -79,6 +79,16 @@
     %elif 'creator' in twitter_card:
     <meta name="twitter:creator" content="${twitter_card['creator']}">
     %endif
+    <meta property="twitter:title" content="${post.title()[:70]|h}">
+    %if post.description():
+    <meta property="twitter:description" content="${post.description()[:200]|h}">
+    %else:
+    <meta property="twitter:description" content="${post.text(strip_html=True)[:200]|h}">
+    %endif
+    %if post.previewimage:
+    <meta property="twitter:image" content="${url_replacer(permalink, post.previewimage, lang)}">
+    <meta property="twitter:image:src" content="${url_replacer(permalink, post.previewimage, lang)}">
+    %endif
 %endif
 </%def>
```

Also there is a tool to validate such markup: http://smo.knowem.com/  It also supports G+ metadata.

**Comment by martin-ueding** [comment]
It seems that with the current boot4 theme the URL is a relative one again. So now the card on Twitter does not work again, because it cannot find the image. I presume that the theme has to be changed such that it uses the full URL?

**Comment by Kwpolska** [comment]
@martin-ueding Cannot reproduce with v8.1.1:

```
.. previewimage: /hello.png
<meta property="og:image" content="https://example.com/hello.png">

.. previewimage: hello2.png
<meta property="og:image" content="https://example.com/posts/welcome-to-nikola/hello2.png">
```

(We recommend to use a path starting with `/`.)

**Comment by martin-ueding** [comment]
I have found the issue! In the `conf.py` I had set `SITE_URL = '/'` because the logo would always redirect me to the production site, even when testing locally using `nikola serve`. This meant that without realizing it I would switch to production and get confused as why changes did not show up. Now with the full URL there, it works like a charm.

**Comment by Kwpolska** [comment]
(The behavior you’re referring to was fixed 5 months ago, in 440583c37/#3229/#3355).

**Comment by martin-ueding** [comment]
I also found that this has improved and I am very happy out it! Now everything works just fine. And since the switch to Nikola, I have been blogging so much more 😊.
